File: 0695-max-area-of-island/0695-max-area-of-island.py
class Solution:
    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        row, col, maxArea = len(grid), len(grid[0]), 0
        visited = set()
        def dfs(r, c, grid, visited):
            if( r not in range(row) or c not in range(col)or
                (r,c) in visited or grid[r][c] == 0
            ):
                return 0
            
            visited.add((r, c))
            return (1 + dfs(r + 1, c, grid, visited) + 
            dfs(r - 1, c, grid, visited) +
            dfs(r , c + 1, grid, visited) +
            dfs(r, c - 1, grid, visited))
        
        for r in range(row):
            for c in range(col):
                if grid[r][c] == 1 and (r, c) not in visited:
                    
                    maxArea = max(maxArea, dfs(r,c, grid, visited))
        return maxArea
        

File: 0695-max-area-of-island/0695-max-area-of-island.java
class Solution {
    public int maxAreaOfIsland(int[][] grid) {
        int rows = grid.length;
        int cols = grid[0].length;
        int maxArea = 0;
        for(int i = 0; i < rows; i++){
            for(int j = 0; j < cols; j++){
                if(grid[i][j] == 1){
                    maxArea = Math.max(maxArea, dfs(rows, cols, i, j, grid));
                }
            }
        }
        return maxArea;
    }
    public int dfs(int r, int c, int i, int j, int[][] grid){
        if(i < 0 || i >= r || j < 0 || j >= c || grid[i][j] == 0) return 0;
        grid[i][j] = 0;
        return (1 + dfs(r, c, i, j + 1, grid)
        +dfs(r, c, i + 1, j, grid)
        +dfs(r, c, i - 1, j, grid)
        +dfs(r, c, i, j - 1, grid));
    }
}